File: C#/106. Construct Binary Tree from Inorder and Postorder Traversal.cs
/**
 * Link: https://leetcode.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/
 * Author: Dung Nguyen Tien (Chris)
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    int[] iOrder;
    int[] pOrder;
    
    public TreeNode BuildTree(int[] inorder, int[] postorder) {
        this.iOrder = inorder;
        this.pOrder = postorder;
        
        var root = helper(postorder.Length - 1, 0, inorder.Length - 1);
        
        return root;
    }
    
    public TreeNode helper(int pStart, int iStart, int iEnd) {
        if (pStart < 0 || iStart > iEnd) return null;
        
        var node = new TreeNode(this.pOrder[pStart]);
        var index = Array.IndexOf(this.iOrder, node.val);
        var rightIndex = pStart - 1;
        var leftIndex = pStart - 1 - (iEnd - index); // inEnd - index is the number of right elements
        
        node.right = helper(rightIndex, index + 1, iEnd);
        node.left = helper(leftIndex, iStart, index - 1);
        return node;
    }
}

File: Python/106. Construct Binary Tree from Inorder and Postorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        def construct(iS, iE, pS, pE):
            if pS > pE: return None
            index = inorder.index(postorder[pE])
            node = TreeNode(postorder[pE])
            node.left = construct(iS, index - 1, pS, pS + (index - iS - 1))
            node.right = construct(index + 1, iE, pS + (index - iS), pE - 1)
            return node
        return construct(0, len(inorder) - 1, 0, len(postorder) - 1)
                    